Initialize intra-op threads in JIT thread pool

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19105 Port TH library to ATen/Parallel
* **#19058 Initialize intra-op threads in JIT thread pool**

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

Differential Revision: [D14849624](https://our.internmc.facebook.com/intern/diff/D14849624)

Commit:
Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

dzhulgakov commented on aten/src/ATen/Parallel.cpp:
```diff
@@ -60,4 +60,29 @@ size_t get_num_threads() {
+
+namespace {
+
+std::shared_ptr<TaskThreadPoolBase> createC10ThreadPool(
```
it's a bit sketchy that we ignore some of the parameters here. Maybe put AT_ASSERT that create_new==false?

dzhulgakov replied:
fix this comment?

ilia-cher replied:
will upd

dzhulgakov reviewed (commented):
Btw, what's the general plan - are you planning to consolidate all thread pool initialization in one place (it doesn't really matter whether it's c10 folder or aten folder)

ilia-cher commented:
@dzhulgakov For the inter-op pool the plan is to eventually initialize it in one place including setting the size of the thread pool (which is not addressed in this PR, this PR is only for intra op setting), so that all the implementation details of inter- and intra-op parallelism are hidden behind an interface.

I had to move PT OSS thread pool to aten in this PR because c10 cannot depend on aten (at::init_num_threads() call)

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Move OMP/MKL thread initialization into ATen/Parallel"

Move OMP/MKL thread initialization into ATen/Parallel

Summary:
This PR implements changes described in
https://github.com/pytorch/pytorch/issues/19001

gh-metadata: pytorch pytorch 19011 gh/ilia-cher/1/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

dzhulgakov commented on aten/src/ATen/Parallel.h:
```diff
@@ -141,4 +143,13 @@ inline scalar_t parallel_reduce(
   }
 }
 
+class C10_API PTThreadPool : public c10::ThreadPool {
```
shouldn't it be AT_API ?

it's unfortunate we need to create an extra class here because of c10/aten boundary. Karl and I are working on it :)

ilia-cher replied:
I don't see AT_API used (that is I see it only in older /experimental/ paths), I think we use C10_API or TORCH_API, and in any case all of this is doing the same thing

ilia-cher replied:
or, did you mean that we'd want to reintroduce AT_API? I thought it's better to agree to one thing

Commit:
Update base for Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

ilia-cher commented:
Added asserts

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

Commit:
Update on "Initialize intra-op threads in JIT thread pool"

Initialize intra-op threads in JIT thread pool

Summary:
Similar to initialization in torch initModule and autograd's
Engine::thread_init, we need to call at::init_num_threads when
initializing JIT thread pool threads

gh-metadata: pytorch pytorch 19058 gh/ilia-cher/2/head

ilia-cher commented:
Using CAFFE2_API
